lib/sync.js

```diff
@@ -304,27 +305,29 @@ SyncApi.prototype.sync = function() {
 
     this._running = true;
 
+    if (global.document) {
+        global.document.addEventListener("online", this._onOnline.bind(this), false);
+    }
+
     // We need to do one-off checks before we can begin the /sync loop.
     // These are:
     //   1) We need to get push rules so we can check if events should bing as we get
     //      them from /sync.
     //   2) We need to get/create a filter which we can use for /sync.
 
-    function getPushRules(attempt) {
-        attempt = attempt || 0;
-        attempt += 1;
-
+    function getPushRules() {
         client.getPushRules().done(function(result) {
             debuglog("Got push rules");
             client.pushRules = result;
             getFilter(); // Now get the filter
-        }, retryHandler(attempt, getPushRules));
+        }, function(err) {
+            self._onConnectionReturned = getPushRules;
```

Comment:
can't help feeling that this would be more elegantly achieved by waiting on a promise which resolves when the connection returns. ymmv though; happy for you to stick with this if you prefer.